### DISCUSSION TITLE
Dates with and without Time

### DISCUSSION THREAD

**Discussion by saumier** [culturecreates/artsdata-planet-lavitrine#6] [Modelling discussions] [answered]
Each Event has startDate and EndDate. However, the datatype can be xsd:dateTime or xsd:date. This works with JSON-LD but is not very JSON friendly. 

I propose the following mapping:
1. When the date and time are known --> call the property **startDateTime** or **endDateTime**
2. When only the date (2023-11-01) is known --> call the property **startDate** or **endDate**

The goal here is to return valid data.  

The only other option I can think of is to always return startDateTime (xsd:dateTime) even when the time is not specified and pad the time using midnight or 23:59 or something similar that is hard-coded and agreed upon with the developers.

**Question**: which convention do you prefer
1. different properties like startDateTime (xsd:dateTime) and startDate (xsd:date)
2. padding xsd:dateTime with 23:59

**Comment by saumier** [comment]
@christianroy FYI I updated this discussion text.

**Reply by christianroy** [reply]
@saumier Est-ce qu'on pourrait faire comme Schema.org pour Event, et simplement garder la propriété `startDate` et y mettre soit une date, soit une date + heure quand on l'a? 

Exemple:

```
"startDate": "2023-09-19"
```

et 

```
"startDate": "2023-09-19T23:30:00Z"
```

seraient tous les deux valides?

Sinon, allons-y avec l'option 1 (startDate et startDateTime), je n'aime pas trop l'idée du padding avec une valeur arbritaire...

cc. @GenFab28

**Reply by saumier** [reply]
@christianroy In schema.org the range for `startDate` is either schema:Date or schema:DateTime as you mention. However this causes a significant problem when using the data because the datatype has to be inferred by parsing the string.  A comparison like `startDate > now()` will not work if the datatype is not a dateTime.  Also, a DateTime cannot be converted to a Date simply be chopping off the time part. A dateTime is an absolute point in time on a continuous timeline (think unix seconds), whereas the same moment in time can be 2 different Dates. For example: `Today 10PM ET 'is after' Tomorrow 1AM UTC` is true, but remove the time part and time zone and you get `Today 'is after' Tomorrow` which is false.  For this reason I find it very important to know the datatype of the startDate. What do you think?

**Reply by christianroy** [reply]
D'accord avec tout ça... J'imaginais effectivement que le consommateur de la donnée devrait inférer la nature de ce qu'il reçoit à partir de la string et choisir en fonction de son contexte comment l'interpréter. Mais c'est vrai que rendu là, entre un `if` sur la longueur de la string, ou un `if` sur le nom de la propriété, le deuxième cas est probablement plus propre. Donc on revient à ton hypothèse 1 (en réitérant que je pense que de padder avec une valeur arbitraire comme 23:59 serait le moins bon scénario).

cc, @GenFab28

**Comment by saumier** [comment]
see issue https://github.com/culturecreates/artsdata-planet-lavitrine/issues/14

**Comment by saumier** [comment, marked as answer]
When both date and time are known (2023-11-01T20:00:00-04:00) --> use the property **startDateTime** and **endDateTime**

When only the date is known (2023-11-01) --> use the property **startDate** or **endDate**
